api/saml/configuration.py

```diff
@@ -0,0 +1,376 @@
+import pickle
+
+from onelogin.saml2.settings import OneLogin_Saml2_Settings
+
+from api.saml.exceptions import SAMLError
+from api.saml.metadata import ServiceProviderMetadata, IdentityProviderMetadata
+from core.model import ConfigurationSetting
+from flask_babel import lazy_gettext as _
+
+
+class SAMLConfigurationSerializingError(SAMLError):
+    """Raised in the case of any errors during configuration serializing"""
+
+
+class SAMLConfigurationSerializer(object):
+    """Serializes and deserializes values as library's configuration settings"""
+
+    def __init__(self, integration):
+        """Initializes a new instance of SAMLConfigurationSerializer class
+
+        :param integration: External integration
+        :type integration: ExternalIntegration
+        """
+        self._integration = integration
+
+    def serialize(self, setting_name, value):
+        """Serializes the value as a pickled library's configuration setting
+
+        :param setting_name: Name of the library's configuration setting
+        :type setting_name: string
+
+        :param value: Value to be serialized
+        :type value: Any
+        """
+        ConfigurationSetting.for_externalintegration(
+            setting_name,
+            self._integration).value = value
+
+    def deserialize(self, setting_name):
+        """Deserializes and returns the library's configuration setting
+
+        :param setting_name: Name of the library's configuration setting
+        :type setting_name: string
+
+        :return: Any
+        """
+        value = ConfigurationSetting.for_externalintegration(
+            setting_name,
+            self._integration).value
+
+        return value
+
+
+class SAMLMetadataSerializer(SAMLConfigurationSerializer):
+    """Serializes and deserializes values as pickled library's configuration settings"""
+
+    def __init__(self, integration):
+        """Initializes a new instance of SAMLMetadataSerializer class
+
+        :param integration: External integration
+        :type integration: ExternalIntegration
+        """
+        super(SAMLMetadataSerializer, self).__init__(integration)
+
+    def serialize(self, setting_name, value):
+        """Serializes the value as a pickled library's configuration setting
+
+        :param setting_name: Name of the library's configuration setting
+        :type setting_name: string
+
+        :param value: Value to be serialized (should be picklable)
+        :type value: Any
+        """
+        if not value:
+            raise ValueError('Value must be non-empty')
+
+        serialized_value = pickle.dumps(value)
```

Comment:
Up to this point we've been using `json.dumps` to convert complex data structures to strings for storage in `configurationsettings.value`. I'm concerned about using `pickle.dumps` for a couple reasons. First, any kind of inconsistency will make the system harder to understand. Second, configurationsettings.value is defined as a Unicode value. In the database everything is binary, of course, so it's theoretically possible to store a binary value here. And presumably the code you wrote works. But it's another inconsistency and it could also cause problems down the road -- I remember we had some problems with Unicode database fields containing non-Unicode data when we were initially preparing for a Python 3 migration.

As I go further into the file, I'll understand more about what this value looks like and my opinion may change, but right now I think you should serialize this to/from JSON rather than using pickle. See https://github.com/NYPL-Simplified/circulation/pull/1441/ for some work I did recently with JSON serialization of a complex value stored in `configurationsettings.value`

Reply:
Another problem I've had in the past with long-term storage of data as pickle objects is, pickle objects embed the Python class and module structure that created them. This means a change to the product that does nothing but refactor code can break existing installations. It's also very difficult to migrate data that's in pickle format. (It's not super easy to migrate JSON data stored in a database, but pickle data is even worse.)

So -- still not having gotten all the way through this file -- I now think that if the IdP configuration data comes in as XML, it should be stored in the database as XML, and parsed into Python objects as necessary. Storing it in any other format is introducing an additional layer at which things can go wrong.

If

Reply:
You're absolutely right regarding pickle - it's not safe. 
In the beginning, I was thinking about not introducing new settings and dynamically parsing XML metadata but I then thought that it might affect the performance: IdP configuration might contain multiple IdPs and parsing it might take time.

Reply:
Performance shouldn't be an issue as the XML file will only be parsed on startup and when an admin changes the site configuration. (That said, if parsing the XML file takes 30 seconds, that's a long time to wait whenever someone changes the site configuration.)

Reply:
Actually, those settings are used during the authentication process. We have to specify SP and IdP metadata for the python-saml library to work.
So every time, a client tries to authenticate by calling `saml_authenticate` endpoint and when IdP sends a SAML assertion by calling `saml_callback` we will have to parse them

Reply:
The `SAMLAuthenticationProvider` object will be instantiated on startup and whenever site configuration is reloaded. If you can create the `SAMLConfiguration` object inside the constructor, you should be able to refer to it as needed without parsing the document every time.

Reply:
Just don't store a reference to the database connection or any database objects inside the `SAMLAuthenticationProvider`; that can cause problems in a multithreaded environment.

Reply:
Sounds great, I'll implement it this way, thank you!

Reply:
Got rid of `SAMLMetadataSerializer` and didn't even need to use `json.dumps`/`json.loads` because I don't store Python objects anymore XML only which is translated into Python objects dynamically